### PR TITLE
skips bad addresses before sending packets

### PR DESCRIPTION
**Pull request by behzadnouri**
#### Problem
bad ip addresses in nodes contact-info.

#### Summary of Changes
* partially emulating current nightly `std::net::IpAddr::is_global` to filter addresses before sending packets.
https://doc.rust-lang.org/std/net/enum.IpAddr.html#method.is_global

### CODE REVIEW

File: streamer/src/socket.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    true
+}
+
+#[cfg(not(debug_assertions))]
+pub fn is_global(addr: &SocketAddr) -> bool {
+    use std::net::IpAddr;
+
+    !match addr.ip() {
+        IpAddr::V4(addr) => {
+            addr.is_private()
```
Comment (sakridge) [inline streamer/src/socket.rs:16]: `is_private` should be ok, no?

Comment (behzadnouri) [reply streamer/src/socket.rs]: done

**Review by t-nelson — APPROVED** [review]
lgtm.  Thanks for the quick turnaround!

**Comment by t-nelson** [comment]
This should be safe to backport to 1.6 and 1.7, right?

**Comment by behzadnouri** [comment]
> This should be safe to backport to 1.6 and 1.7, right?

yeah, I think so

**Comment by ryoqun** [comment]
@behzadnouri it looks like this broke system-performance-tests... should we change to use global ips always for internal tests as well or could we introduce some flags to allow private addresses?

**Comment by CriesofCarrots** [comment]
@behzadnouri I also had a report that this has broken the `--gossip-host` parameter for `solana-test-validator` running in a docker container. I plan to repro and get back with more details, but it would be great to fix that use-case as well

**Comment by behzadnouri** [comment]
@ryoqun @CriesofCarrots thanks for reporting the issues. let me see if I can find a good fix. I hope we can avoid using flags as that would pollute too many places.

**Comment by behzadnouri** [comment]
One solution might be to add a new feature to `streamer/Cargo.toml`:
```
[features]
allow_private_addr = []
```
and build with this feature for tests, as in: https://github.com/behzadnouri/solana/commit/bad-addr
Not sure if it will work though.

File: streamer/src/socket.rs (hunk 1 above)
Comment (sakridge) [reply streamer/src/socket.rs]: I meant that, `is_private()` should be allowed, but the other ranges should not.

**Comment by sakridge** [comment]
> @behzadnouri it looks like this broke system-performance-tests... should we change to use global ips always for internal tests as well or could we introduce some flags to allow private addresses?

We don't want to use global IPs, it causes outbound traffic which you need to pay for. We should preserve the ability to operate on a local cluster with the 192.168.x.x/10.0.x.x ranges.

**Comment by behzadnouri** [comment]
I am testing https://github.com/solana-labs/solana/pull/18806
Lets see if it works.

**Comment by mvines** [comment]
Actually I think we need that runtime flag.  `is_private()` breaks our ability to:
1. Run a cluster over a private IP range in GCP (ie, avoid hitting the public IPs and 💸 due to egress charges), and otherwise run any private IP cluster in general
2. Breaks `solana-test-validator` from running inside Docker

**Comment by mvines** [comment]
One possible approach: if your advertised gossip address is public, then apply the logic here

**Comment by behzadnouri** [comment]
Lets move the discussion/comments to https://github.com/solana-labs/solana/pull/18806 which adds a _development only_ build feature.

I had a successful test run of ` system-performance-tests` with that commit:
https://buildkite.com/solana-labs/system-performance-tests/builds/5075

**Pull request merged**